coralproject/talk
# [CORL-2222/2223/2225] Add story archiving features to Coral

## What does this PR do?

Adds the ability to archive stories to a secondary mongo instance.

Adds the mutations to archive stories and also adds the auto-archiving job features to let stories archive as they age and pass a certain age.

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds two mutations to `archiveStories` and `unarchiveStories` along with their input's and payloads.

## How do I test this PR?

- Spin up another mongo instance to be your archiving Db
```
docker run -d -p 27018:27017 --restart always --name mongo-archive mongo:4.2
```
- Set the environment variable for this new archiving URI
```
MONGODB_ARCHIVE_URI=mongodb://127.0.0.1:27018/coral
```
- Find a story to archive and use the following mutation to do so:
```
mutation($storyIDs: [ID!]!, $clientMutationId: String!) {
  archiveStories(input: {
    storyIDs: $storyIDs,
    clientMutationId: $clientMutationId
  }) {
    stories {
      id
      isArchiving
      isArchived
    }
  }
}

// Query variables
{
	"storyIDs": [
		"<STORY_ID_YOU_WANT_TO_ARCHIVE>"
	],
	"clientMutationId": "0"
}
```
- Go check the story in stream and see that it is closed, archived, and is completely read-only
 
## How do we deploy this PR?

- If a client is not using archiving, you don't need to do anything.

- If you want to only use archiving and only use manual archive mutations on a client, you need to set the archive URI env variable on their regular and job deploy yaml:

```
MONGODB_ARCHIVE_URI
```
(if this is not set, it will default to the main mongodb URI.)

- If you want to use auto-archiving job features, you need to set the following on the job deploy yaml (or whichever instance will do archiving jobs).

```
ENABLE_AUTO_ARCHIVING <-- set to true to turn archive jobs on
AUTO_ARCHIVE_OLDER_THAN <-- optional, has a default
AUTO_ARCHIVING_BATCH_SIZE <-- optional, has a default
```
